<a href="https://colab.research.google.com/github/DiegoEspinozaoss/Natural_Language_Processing_CC6205-1/blob/main/Tarea_1_IR.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Tarea 1: Introducción, Modelos de Espacio Vectorial, Recuperación de Información y Modelos de Lenguaje

**Procesamiento de Lenguaje Natural (CC6205-1 - Otoño 2025)**

## Tarjeta de Identificación

- **Nombre(s):** ```Diego Espinoza, Juan Miño, Emilio Torres```
- **Fecha límite de entrega 📆:** 15 de abril de 2025
- **Tiempo estimado de dedicación:** 4 horas

---

## 📌 Instrucciones

¡Bienvenid@s a la primera tarea del curso de *Natural Language Processing* (NLP)!

El objetivo de esta tarea es evaluar los conceptos teóricos de las primeras semanas de clases, centrándose en:

- **Recuperación de Información (IR)**
- **Modelos de Espacio Vectorial**
- **Modelos de Lenguaje**

Si aún no has revisado el contenido correspondiente, se recomienda consultar las referencias disponibles al final del documento.

### 📢 Consideraciones generales

✅ La tarea debe realizarse en **grupos de hasta 3 personas**.

✅ La entrega debe realizarse a través de **U-Cursos**, a más tardar en la fecha estipulada. **No se aceptarán entregas atrasadas.**

✅ El formato de entrega es este mismo **Jupyter Notebook**.

✅ Su código será ejecutado al momento de la revisión. **Verifiquen que no tenga errores de compilación.**

✅ Es obligatorio completar la **Tarjeta de Identificación**. **No se asignará nota sin ella.**

---

## 📚 Material de Referencia

### 📄 Diapositivas del curso

- [Introducción al curso](https://github.com/dccuchile/CC6205/blob/master/slides/NLP-introduction.pdf)
- [Modelos de Espacio Vectorial y Recuperación de Información](https://github.com/dccuchile/CC6205/blob/master/slides/NLP-IR.pdf)
- [Modelos Probabilísticos de Lenguaje](https://github.com/dccuchile/CC6205/blob/master/slides/NLP-PLM.pdf)

### 📺 Videos del curso

- **Introducción**: [Parte 1](https://www.youtube.com/watch?v=HEKTNOttGvU) | [Parte 2](https://www.youtube.com/watch?v=P8cwnI-f-Kg)
- **Recuperación de Información**: [Parte 1](https://www.youtube.com/watch?v=FXIVClF370w&list=PLppKo85eGXiXIh54H_qz48yHPHeNVJqBi&index=3) | [Parte 2](https://www.youtube.com/watch?v=f8nG1EMmPZk&list=PLppKo85eGXiXIh54H_qz48yHPHeNVJqBi&index=3)
- **Modelos Probabilísticos de Lenguaje**: [Parte 1](https://www.youtube.com/watch?v=9E2jJ6kcb4Y&list=PLppKo85eGXiXIh54H_qz48yHPHeNVJqBi&index=3) | [Parte 2](https://www.youtube.com/watch?v=ZWqbEQXLra0&list=PLppKo85eGXiXIh54H_qz48yHPHeNVJqBi&index=5) | [Parte 3](https://www.youtube.com/watch?v=tsumFqwFlaA&list=PLppKo85eGXiXIh54H_qz48yHPHeNVJqBi&index=6) | [Parte 4](https://www.youtube.com/watch?v=s3TWdv4sqkg&list=PLppKo85eGXiXIh54H_qz48yHPHeNVJq)

---

📌 **Recuerda:** La claridad y organización en la entrega son clave para una mejor evaluación. Incluir analisis apropiado. ¡Mucho éxito! 🚀

## P1. Tokenización

En el primer ejercicio veremos la dificultad de tokenizar textos no estructurados, destacando la importancia de tener librerías que realicen este trabajo.

In [ ]:
# En caso de desarrollar la tarea desde colab, con el siguiente código podemos cargar los archivos desde drive:

try:
    from google.colab import drive

    drive.mount("/content/drive", force_remount=True)
    path = '/content/drive/MyDrive/nlp/oh_algoritmo.txt'
except:
    print('Ignorando conexión drive-colab')

Ignorando conexión drive-colab


Ejecute el código a continuación para cargar el ejemplo. Recuerde realizar la modificación al directorio en caso que el archivo no se encuentre en el mismo directorio del Jupyter Notebook


In [5]:
try:
    # Abre el archivo en modo lectura ("r")
    with open("oh_algoritmo.txt", "r") as archivo:
        # Lee el contenido del archivo
        texto = archivo.read()
        # Imprime el contenido
        print(texto)
except FileNotFoundError:
    print("El archivo no se encuentra.")
except Exception as e:
    print("Ocurrió un error:", e)

[Letra de "¡Oh, Algoritmo!" ft. Nora Erez]

[Refrán: Jorge Drexler]
¿Quién quiere que yo quiera lo que creo que quiero?
¿Quién quiere que yo quiera lo que creo que quiero?
¿Quién quiere que yo quiera lo que creo que quiero?
¿Quién quiere que yo quiera lo que creo que quiero?
¿Quién quiere que yo quiera lo que creo que quiero?
¿Quién quiere que yo quiera lo que creo que quiero?

[Estribillo: Jorge Drexler]
Dime qué debo cantar
Oh, algoritmo
Sé que lo sabes mejor
Incluso que yo mismo

[Verso 1: Nora Erez]
Wait, what's that money that you spent?
What's that sitting on your plate?
Do you want what you've been fed?
Are you the fish or bait?
Mmm, I'm on the top of the roof and I feel like a jail
Rather not pay the bail
To dangerous people with blood on their faces
So I'm sharing a cell with the masses
The underground always strive for the main
Streaming like Grande's big-ass ring
Screaming: I'll write you out my will
Conscious is free, but not the will
Conscious is free, but 

Fuente: https://genius.com/Jorge-drexler-oh-algoritmo-lyrics

### Pregunta 1.a (0.25 puntos)

Diseñe una función **`get_tokens()`** que reciba un texto y entregue una lista con sus tokens. Es libre de elegir la forma de tokenizar mientras no utilice librerías con tokenizadores ya implementados. Puede utilizar la librería **re** importada para trabajar símbolos. Explique su razonamiento.


In [6]:
import re

In [9]:
def get_tokens(texto):
    tokens = re.findall(r'\b\w+\b', texto)
    return tokens
    pass

In [8]:
tokens = get_tokens(texto)
tokens

['Letra',
 'de',
 'Â',
 'Oh',
 'Algoritmo',
 'ft',
 'Nora',
 'Erez',
 'RefrÃ',
 'n',
 'Jorge',
 'Drexler',
 'Â',
 'QuiÃ',
 'n',
 'quiere',
 'que',
 'yo',
 'quiera',
 'lo',
 'que',
 'creo',
 'que',
 'quiero',
 'Â',
 'QuiÃ',
 'n',
 'quiere',
 'que',
 'yo',
 'quiera',
 'lo',
 'que',
 'creo',
 'que',
 'quiero',
 'Â',
 'QuiÃ',
 'n',
 'quiere',
 'que',
 'yo',
 'quiera',
 'lo',
 'que',
 'creo',
 'que',
 'quiero',
 'Â',
 'QuiÃ',
 'n',
 'quiere',
 'que',
 'yo',
 'quiera',
 'lo',
 'que',
 'creo',
 'que',
 'quiero',
 'Â',
 'QuiÃ',
 'n',
 'quiere',
 'que',
 'yo',
 'quiera',
 'lo',
 'que',
 'creo',
 'que',
 'quiero',
 'Â',
 'QuiÃ',
 'n',
 'quiere',
 'que',
 'yo',
 'quiera',
 'lo',
 'que',
 'creo',
 'que',
 'quiero',
 'Estribillo',
 'Jorge',
 'Drexler',
 'Dime',
 'quÃ',
 'debo',
 'cantar',
 'Oh',
 'algoritmo',
 'SÃ',
 'que',
 'lo',
 'sabes',
 'mejor',
 'Incluso',
 'que',
 'yo',
 'mismo',
 'Verso',
 '1',
 'Nora',
 'Erez',
 'Wait',
 'what',
 's',
 'that',
 'money',
 'that',
 'you',
 'spent',
 'What',


### Pregunta 1.b (0.25 puntos)
Explique su implementación aquí:
La implementación ... Esta librería "re" cuyo método findall(r'\b\w+\b',texto) es tal que; \b es un límite de palabra (asegura que no tomemos fragmentos de palabras como partes de los tokens), \w+ captura una o más letras, números o guiones bajos (es decir, palabras). Esto asegura que cada token sea una secuencia de caracteres alfanuméricos. Esta expresión regular no captura signos de puntuación ni espacios, solo las palabras. Por ejemplo, si tenemos 

>movie_str = "the evening shows start at 7:00pm and 10:15pm"

>matches = re.findall(r"([\d:,.]+)(am|pm)?", movie_str)

devolverá lo siguiente: 

>[("7:00", "pm"), ("10:15", "pm")]. 

Entonces la sintaxis general de re.findall es la [siguiente](https://www.codecademy.com/resources/docs/python/regex/findall): re.findall(<pattern>, string), donde pattern vendría siendo el patrón que puede incluir un string, un código de clase de carácter como /w, /s o /d, o bien un símbolo regular como $, | y ^.

Implementación con la libreria NLTK

In [ ]:
from nltk.tokenize import wordpunct_tokenize
nltk_tokens = wordpunct_tokenize(texto)
nltk_tokens

['[',
 'Letra',
 'de',
 '"¡',
 'Oh',
 ',',
 'Algoritmo',
 '!"',
 'ft',
 '.',
 'Nora',
 'Erez',
 ']',
 '[',
 'Refrán',
 ':',
 'Jorge',
 'Drexler',
 ']',
 '¿',
 'Quién',
 'quiere',
 'que',
 'yo',
 'quiera',
 'lo',
 'que',
 'creo',
 'que',
 'quiero',
 '?',
 '¿',
 'Quién',
 'quiere',
 'que',
 'yo',
 'quiera',
 'lo',
 'que',
 'creo',
 'que',
 'quiero',
 '?',
 '¿',
 'Quién',
 'quiere',
 'que',
 'yo',
 'quiera',
 'lo',
 'que',
 'creo',
 'que',
 'quiero',
 '?',
 '¿',
 'Quién',
 'quiere',
 'que',
 'yo',
 'quiera',
 'lo',
 'que',
 'creo',
 'que',
 'quiero',
 '?',
 '¿',
 'Quién',
 'quiere',
 'que',
 'yo',
 'quiera',
 'lo',
 'que',
 'creo',
 'que',
 'quiero',
 '?',
 '¿',
 'Quién',
 'quiere',
 'que',
 'yo',
 'quiera',
 'lo',
 'que',
 'creo',
 'que',
 'quiero',
 '?',
 '[',
 'Estribillo',
 ':',
 'Jorge',
 'Drexler',
 ']',
 'Dime',
 'qué',
 'debo',
 'cantar',
 'Oh',
 ',',
 'algoritmo',
 'Sé',
 'que',
 'lo',
 'sabes',
 'mejor',
 'Incluso',
 'que',
 'yo',
 'mismo',
 '[',
 'Verso',
 '1',
 ':',
 'Nora',
 

### Pregunta 1.c (0.5 puntos)
¿Qué diferencias y similitudes encontrase al comparar la función de tokenización creada manualmente por ti contra la implementación de NLTK, al tokenizar la letra de la canción "Oh, algoritmo"?

>Una diferencia que encontré es que en el caso de la función de tokenización creada a mano no se aceptan tokens como simbolos tales como "?", ")", "]", etc., mientras que en el NLTK sí. Una similitud es que parecen respetar el orden del texto original, es decir, parecen guardar elementos secuencialmente en orden de aparición a medida que van leyendo el teto. Por otro lado, en la función que definí yo es necesario especificar los caracteres que se consideran válidos, mientras que en la NLTK no. Así mismo, la función mía aceptaba letras como â o Â, mientras que la NLTK no, quizás debido a su enfoque. 

## P2. Stemming y Stopwords

En esta sección debera implementar funciones de stemming y stopwords basado en lo visto en clase. En la siguiente celda tiene el corpus que usara en esta sección:

In [ ]:
# Corpus en español
corpus_espanol = [
    "¿Quién quiere que yo quiera lo que creo que quiero?",
    "Dime qué debo cantar",
    "Sé que lo sabes mejor"
]

# Corpus en inglés
corpus_ingles = [
    "What's that sitting on your plate?",
    "Do you want what you've been fed?",
    "Are you the fish or bait?"
]

### Pregunta 2.a (0.5 puntos)
Implemente una función **`get_vocab()`** que extraiga los tokens de un corpus. Puede utilizar la función de la sección anterior.

In [ ]:
def get_vocab(corpus):
  ### Aquí inicia tu código ###
  pass
  ### Aquí termina tu código ###

In [ ]:
vocab_espanol = get_vocab(corpus_espanol)
vocab_espanol

['yo',
 'qué',
 'Quién',
 'quiera',
 'Dime',
 'que',
 'Sé',
 'sabes',
 'cantar',
 'quiere',
 'lo',
 'mejor',
 'debo',
 'quiero',
 'creo']

Resultado esperado (el orden puede variar):
```
['yo', 'debo', 'creo', 'Dime', 'lo', 'cantar', 'mejor', 'Sé', 'que', 'quiere', 'quiero', 'sabes', 'Quién', 'quiera', 'qué']
```

In [ ]:
vocab_ingles = get_vocab(corpus_ingles)
vocab_ingles

['want',
 'Are',
 'bait',
 'your',
 'been',
 's',
 'you',
 've',
 'sitting',
 'Do',
 'the',
 'fed',
 'on',
 'that',
 'fish',
 'plate',
 'or',
 'What',
 'what']

Resultado esperado:
```
['fed', 'been', 'or', 'want', 'plate', 'the', 've', 'your', 's', 'you', 'what', 'Are', 'bait', 'What', 'fish', 'that', 'sitting', 'Do', 'on']
```

### Pregunta 2.b (0.5 puntos)
Ahora diseñe reglas que usted estime convenientes tanto de **Stemming** como de **Stopwords**. Implemente una función que reciba una lista con los elementos del vocabulario, le aplique sus reglas y devuelva el vocabulario preprocesado. Explique las reglas de stemming y elección de stopwords:

    Explique sus reglas aquí:


In [ ]:
def pre_processing(vocabulario, idioma):
    ### Aquí inicia tu código ###
    pass
    ### Aquí termina tu código ##

In [ ]:
# Aplicar preprocesamiento a los vocabularios de ejemplo con NLTK
vocab_procesado_espanol = pre_processing(vocab_espanol, 'espanol')
vocab_procesado_ingles = pre_processing(vocab_ingles, 'ingles')

# Mostrar resultados
print("Vocabulario procesado en español:", vocab_procesado_espanol, "\n")
print("Vocabulario procesado en inglés:", vocab_procesado_ingles, "\n")

Vocabulario procesado en español: ['Quién', 'quier', 'Dim', 'sabe', 'cant', 'quier', 'mejor', 'quier', 'cre'] 

Vocabulario procesado en inglés: ['want', 'bait', 'sitt', 'fish', 'plate'] 



## P3. Bag of Words (0.5 puntos)
Considere el siguiente corpus, donde cada elemento del arreglo representa un documento:


In [ ]:
d0 = 'El pájaro come semillas'
d1 = 'El pájaro se despierta y canta'
d2 = 'El pájaro canta y come semillas'
d3 = 'El pez come y nada en el agua'
d4 = 'El pez empieza a nadar'
d5 = 'El pez come alimento'
corpus = [d0, d1, d2, d3, d4, d5]

El objetivo da las siguientes secciones es determinar cuáles de  los documentos entregados son los más similares entre sí. Para ello utilizaremos la técnica **TF-IDF**.

Como los algoritmos de Machine Learning no comprenden el texto en lenguaje natural, estos documentos deben ser convertidos a vectores numéricos. La representación más simple vista en clases es la de **Bag of Words**, método mediante el cual se cuentan las apariciones de cada palabra en cada uno de los documentos entregados.

Implemente la función **`bag_of_words()`**, que recibe como input un arreglo de documentos y devuelve un dataframe de pandas con la representación Bag of Words de los documentos entregados. En esta representación las columnas son el vocabulario y las filas representan las apariciones de cada una de las palabras en los documentos. En otras palabras, cada fila representa el BoW de un documento.

***Disclaimer: el orden de los resultados pueden variar.***


Por ejemplo para el siguiente corpus:

```
corpus = ['El perro ladra', 'El perro come']
```

Debiese entregarnos lo siguiente:


|   | el | perro | ladra | come |
|---|----|-------|------|-------|
| 0 | 1  | 1     | 1    | 0     |
| 1 | 1  | 1     | 0    | 1     |




In [ ]:
import pandas as pd

Implementar función `bag_of_words()`

In [ ]:
def bag_of_words(corpus):
    ### Aquí inicia tu código ###
    pass
    ### Aquí termina tu código ##

In [ ]:
dataset_bow = bag_of_words(corpus)
dataset_bow

El  pájaro  despierta  el  come  a  nadar  se  en  y  alimento  semillas  \
d0   1       1          0   0     1  0      0   0   0  0         0         1   
d1   1       1          1   0     0  0      0   1   0  1         0         0   
d2   1       1          0   0     1  0      0   0   0  1         0         1   
d3   1       0          0   1     1  0      0   0   1  1         0         0   
d4   1       0          0   0     0  1      1   0   0  0         0         0   
d5   1       0          0   0     1  0      0   0   0  0         1         0   

    pez  empieza  canta  agua  nada  
d0    0        0      0     0     0  
d1    0        0      1     0     0  
d2    0        0      1     0     0  
d3    1        0      0     1     1  
d4    1        1      0     0     0  
d5    1        0      0     0     0

Solución esperada:

|    |   El |   pájaro |   despierta |   el |   come |   a |   nadar |   se |   en |   y |   alimento |   semillas |   pez |   empieza |   canta |   agua |   nada |
|:---|-----:|---------:|------------:|-----:|-------:|----:|--------:|-----:|-----:|----:|-----------:|-----------:|------:|----------:|--------:|-------:|-------:|
| d0 |    1 |        1 |           0 |    0 |      1 |   0 |       0 |    0 |    0 |   0 |          0 |          1 |     0 |         0 |       0 |      0 |      0 |
| d1 |    1 |        1 |           1 |    0 |      0 |   0 |       0 |    1 |    0 |   1 |          0 |          0 |     0 |         0 |       1 |      0 |      0 |
| d2 |    1 |        1 |           0 |    0 |      1 |   0 |       0 |    0 |    0 |   1 |          0 |          1 |     0 |         0 |       1 |      0 |      0 |
| d3 |    1 |        0 |           0 |    1 |      1 |   0 |       0 |    0 |    1 |   1 |          0 |          0 |     1 |         0 |       0 |      1 |      1 |
| d4 |    1 |        0 |           0 |    0 |      0 |   1 |       1 |    0 |    0 |   0 |          0 |          0 |     1 |         1 |       0 |      0 |      0 |
| d5 |    1 |        0 |           0 |    0 |      1 |   0 |       0 |    0 |    0 |   0 |          1 |          0 |     1 |         0 |       0 |      0 |      0 |

## P4. TF-IDF

### 4.a TF (0.25 puntos)

Ahora debemos usar el dataframe del ejercicio anterior para calcular la matriz de TF normalizada por la máxima frecuencia $\max_i({\text{tf}_{i,j}})$, donde
$i$ corresponde al índice de las filas (BoW) y $j$ al de las columnas (palabras). Es decir, dividir cada BoW sobre la cantidad de veces de la palabra que aparezca más veces en ese vector.


$$\text{nft}_{i,j} = \frac{\text{tf}_{i,j}}{\max_i({\text{tf}_{i,j})}}$$

Implemente la función `calc_tf(dataset_bow)`, que entrega la matriz de TF normalizada del BoW del dataset.

In [ ]:
def calc_tf(dataset_bow):
    ### Aquí inicia tu código ###
    pass
    ### Aquí termina tu código ##

In [ ]:
tf = calc_tf(dataset_bow)
tf

El  pájaro  despierta   el  come    a  nadar   se   en    y  alimento  \
d0  1.0     1.0        0.0  0.0   1.0  0.0    0.0  0.0  0.0  0.0       0.0   
d1  1.0     1.0        1.0  0.0   0.0  0.0    0.0  1.0  0.0  1.0       0.0   
d2  1.0     1.0        0.0  0.0   1.0  0.0    0.0  0.0  0.0  1.0       0.0   
d3  1.0     0.0        0.0  1.0   1.0  0.0    0.0  0.0  1.0  1.0       0.0   
d4  1.0     0.0        0.0  0.0   0.0  1.0    1.0  0.0  0.0  0.0       0.0   
d5  1.0     0.0        0.0  0.0   1.0  0.0    0.0  0.0  0.0  0.0       1.0   

    semillas  pez  empieza  canta  agua  nada  
d0       1.0  0.0      0.0    0.0   0.0   0.0  
d1       0.0  0.0      0.0    1.0   0.0   0.0  
d2       1.0  0.0      0.0    1.0   0.0   0.0  
d3       0.0  1.0      0.0    0.0   1.0   1.0  
d4       0.0  1.0      1.0    0.0   0.0   0.0  
d5       0.0  1.0      0.0    0.0   0.0   0.0

Solución esperada:

|    |   El |   pájaro |   despierta |   el |   come |   a |   nadar |   se |   en |   y |   alimento |   semillas |   pez |   empieza |   canta |   agua |   nada |
|:---|-----:|---------:|------------:|-----:|-------:|----:|--------:|-----:|-----:|----:|-----------:|-----------:|------:|----------:|--------:|-------:|-------:|
| d0 |    1 |        1 |           0 |    0 |      1 |   0 |       0 |    0 |    0 |   0 |          0 |          1 |     0 |         0 |       0 |      0 |      0 |
| d1 |    1 |        1 |           1 |    0 |      0 |   0 |       0 |    1 |    0 |   1 |          0 |          0 |     0 |         0 |       1 |      0 |      0 |
| d2 |    1 |        1 |           0 |    0 |      1 |   0 |       0 |    0 |    0 |   1 |          0 |          1 |     0 |         0 |       1 |      0 |      0 |
| d3 |    1 |        0 |           0 |    1 |      1 |   0 |       0 |    0 |    1 |   1 |          0 |          0 |     1 |         0 |       0 |      1 |      1 |
| d4 |    1 |        0 |           0 |    0 |      0 |   1 |       1 |    0 |    0 |   0 |          0 |          0 |     1 |         1 |       0 |      0 |      0 |
| d5 |    1 |        0 |           0 |    0 |      1 |   0 |       0 |    0 |    0 |   0 |          1 |          0 |     1 |         0 |       0 |      0 |      0 |

### 4.b IDF (0.5 puntos)

Implementar `calc_idf(dataset_bow)`. Ésta debe retornar un diccionario en donde las llaves sean las palabras y los valores sean el cálculo de cada idf por palabra.

Recordar que $\text{idf}_{t_i} = \log_{10}\frac{N}{n_i}$ con $N = $ número de documentos y $n_i = $ número de documentos que contienen la palabra $t_i$

In [ ]:
import numpy as np

In [ ]:
def calc_idf(dataset_bow):
    ### Aquí inicia tu código ###
    pass
    ### Aquí termina tu código ##

In [ ]:
idf = calc_idf(dataset_bow)
idf

{'El': 0.0,
 'pájaro': 0.3010299956639812,
 'despierta': 0.7781512503836436,
 'el': 0.7781512503836436,
 'come': 0.17609125905568124,
 'a': 0.7781512503836436,
 'nadar': 0.7781512503836436,
 'se': 0.7781512503836436,
 'en': 0.7781512503836436,
 'y': 0.3010299956639812,
 'alimento': 0.7781512503836436,
 'semillas': 0.47712125471966244,
 'pez': 0.3010299956639812,
 'empieza': 0.7781512503836436,
 'canta': 0.47712125471966244,
 'agua': 0.7781512503836436,
 'nada': 0.7781512503836436}

Solución esperada:
```
{'El': 0.0,
 'pájaro': 0.3010299956639812,
 'despierta': 0.7781512503836436,
 'el': 0.7781512503836436,
 'come': 0.17609125905568124,
 'a': 0.7781512503836436,
 'nadar': 0.7781512503836436,
 'se': 0.7781512503836436,
 'en': 0.7781512503836436,
 'y': 0.3010299956639812,
 'alimento': 0.7781512503836436,
 'semillas': 0.47712125471966244,
 'pez': 0.3010299956639812,
 'empieza': 0.7781512503836436,
 'canta': 0.47712125471966244,
 'agua': 0.7781512503836436,
 'nada': 0.7781512503836436}
 ```

### 4.c TF-IDF (0.25 puntos)
Programe la función `calc_tf_idf(tf, idf)` que entrega el dataframe TF-IDF asociado al dataset que estamos analizando.

In [ ]:
def calc_tf_idf(tf, idf):
    ### Aquí inicia tu código ###
    pass
    ### Aquí termina tu código ##

In [ ]:
tf_idf = calc_tf_idf(tf, idf)
tf_idf

El   pájaro  despierta        el      come         a     nadar        se  \
d0  0.0  0.30103   0.000000  0.000000  0.176091  0.000000  0.000000  0.000000   
d1  0.0  0.30103   0.778151  0.000000  0.000000  0.000000  0.000000  0.778151   
d2  0.0  0.30103   0.000000  0.000000  0.176091  0.000000  0.000000  0.000000   
d3  0.0  0.00000   0.000000  0.778151  0.176091  0.000000  0.000000  0.000000   
d4  0.0  0.00000   0.000000  0.000000  0.000000  0.778151  0.778151  0.000000   
d5  0.0  0.00000   0.000000  0.000000  0.176091  0.000000  0.000000  0.000000   

          en        y  alimento  semillas      pez   empieza     canta  \
d0  0.000000  0.00000  0.000000  0.477121  0.00000  0.000000  0.000000   
d1  0.000000  0.30103  0.000000  0.000000  0.00000  0.000000  0.477121   
d2  0.000000  0.30103  0.000000  0.477121  0.00000  0.000000  0.477121   
d3  0.778151  0.30103  0.000000  0.000000  0.30103  0.000000  0.000000   
d4  0.000000  0.00000  0.000000  0.000000  0.30103  0.778151  0.000000   
d5  0.000000  0.00000  0.778151  0.000000  0.30103  0.000000  0.000000   

        agua      nada  
d0  0.000000  0.000000  
d1  0.000000  0.000000  
d2  0.000000  0.000000  
d3  0.778151  0.778151  
d4  0.000000  0.000000  
d5  0.000000  0.000000

Solución esperada:

|    |   El |   pájaro |   despierta |       el |     come |        a |    nadar |       se |       en |       y |   alimento |   semillas |     pez |   empieza |    canta |     agua |     nada |
|:---|-----:|---------:|------------:|---------:|---------:|---------:|---------:|---------:|---------:|--------:|-----------:|-----------:|--------:|----------:|---------:|---------:|---------:|
| d0 |    0 |  0.30103 |    0        | 0        | 0.176091 | 0        | 0        | 0        | 0        | 0       |   0        |   0.477121 | 0       |  0        | 0        | 0        | 0        |
| d1 |    0 |  0.30103 |    0.778151 | 0        | 0        | 0        | 0        | 0.778151 | 0        | 0.30103 |   0        |   0        | 0       |  0        | 0.477121 | 0        | 0        |
| d2 |    0 |  0.30103 |    0        | 0        | 0.176091 | 0        | 0        | 0        | 0        | 0.30103 |   0        |   0.477121 | 0       |  0        | 0.477121 | 0        | 0        |
| d3 |    0 |  0       |    0        | 0.778151 | 0.176091 | 0        | 0        | 0        | 0.778151 | 0.30103 |   0        |   0        | 0.30103 |  0        | 0        | 0.778151 | 0.778151 |
| d4 |    0 |  0       |    0        | 0        | 0        | 0.778151 | 0.778151 | 0        | 0        | 0       |   0        |   0        | 0.30103 |  0.778151 | 0        | 0        | 0        |
| d5 |    0 |  0       |    0        | 0        | 0.176091 | 0        | 0        | 0        | 0        | 0       |   0.778151 |   0        | 0.30103 |  0        | 0        | 0        | 0        |

## P5. Cosine-similarity (0.25 puntos)
Ahora que tenemos el dataframe de TF-IDF, nos queda calcular la similitud coseno entre todos los vectores. Notar que la matriz resultante será una matriz simétrica.

Implemente la función *cosine_similarity(v1, v2)* que recibe dos vectores (v1 y v2) y calcula la similitud coseno entre ambos. Concluya cuáles son los dos documentos más similares.

In [ ]:
def cosine_similarity(v1, v2):
    ### Aquí inicia tu código ###
    pass
    ### Aquí termina tu código ##

In [ ]:
similarity_matrix = np.zeros((6,6))
for i, v1 in enumerate(tf_idf.index.values):
  for j, v2 in enumerate(tf_idf.index.values):
      similarity = cosine_similarity(tf_idf.loc[v1].values, tf_idf.loc[v2].values)
      similarity_matrix[i][j] = similarity

for i in range(6):
  mask = [k != i for k in range(6)]
  j = np.argmax(similarity_matrix[i][mask])

  print(corpus[i])
  print("> Mas similar:", np.array(corpus)[mask][j])
  print("> Similitud:", similarity_matrix[i][mask][j], "\n")

El pájaro come semillas
> Mas similar: El pájaro canta y come semillas
> Similitud: 0.7233435041520414 

El pájaro se despierta y canta
> Mas similar: El pájaro canta y come semillas
> Similitud: 0.39320101823128945 

El pájaro canta y come semillas
> Mas similar: El pájaro come semillas
> Similitud: 0.7233435041520414 

El pez come y nada en el agua
> Mas similar: El pájaro canta y come semillas
> Similitud: 0.09171890791406168 

El pez empieza a nadar
> Mas similar: El pez come alimento
> Similitud: 0.07695078406752713 

El pez come alimento
> Mas similar: El pez come y nada en el agua
> Similitud: 0.0878790037217323 



Solución esperada:
```
El pájaro come semillas
> Mas similar: El pájaro canta y come semillas
> Similitud: 0.7233435041520414

El pájaro se despierta y canta
> Mas similar: El pájaro canta y come semillas
> Similitud: 0.39320101823128945

El pájaro canta y come semillas
> Mas similar: El pájaro come semillas
> Similitud: 0.7233435041520414

El pez come y nada en el agua
> Mas similar: El pájaro canta y come semillas
> Similitud: 0.09171890791406168

El pez empieza a nadar
> Mas similar: El pez come alimento
> Similitud: 0.07695078406752713

El pez come alimento
> Mas similar: El pez come y nada en el agua
> Similitud: 0.0878790037217323
```

## P6 N-gramas (0.75 punto)

En esta sección debera determinar los n-gramas del la cancion "Oh algoritmo".

### 6.a Corpus de entrenamiento y test (0.25 puntos)

En esta subsección debera definir el conjunto de entrenamiento y test de un corpus. Eliga una particion del 80% y 20% del texto.

In [ ]:
try:
    # Abre el archivo en modo lectura ("r")
    with open("oh_algoritmo.txt", "r") as archivo:
        # Lee el contenido del archivo
        texto = archivo.read()
        # Imprime el contenido
        print(texto)
except FileNotFoundError:
    print("El archivo no se encuentra.")
except Exception as e:
    print("Ocurrió un error:", e)

[Letra de "¡Oh, Algoritmo!" ft. Nora Erez]

[Refrán: Jorge Drexler]
¿Quién quiere que yo quiera lo que creo que quiero?
¿Quién quiere que yo quiera lo que creo que quiero?
¿Quién quiere que yo quiera lo que creo que quiero?
¿Quién quiere que yo quiera lo que creo que quiero?
¿Quién quiere que yo quiera lo que creo que quiero?
¿Quién quiere que yo quiera lo que creo que quiero?

[Estribillo: Jorge Drexler]
Dime qué debo cantar
Oh, algoritmo
Sé que lo sabes mejor
Incluso que yo mismo

[Verso 1: Nora Erez]
Wait, what's that money that you spent?
What's that sitting on your plate?
Do you want what you've been fed?
Are you the fish or bait?
Mmm, I'm on the top of the roof and I feel like a jail
Rather not pay the bail
To dangerous people with blood on their faces
So I'm sharing a cell with the masses
The underground always strive for the main
Streaming like Grande's big-ass ring
Screaming: I'll write you out my will
Conscious is free, but not the will
Conscious is free, but not the will
You

Defina una funcion `get_sentences()` que entregue todas las oraciones del corpus que contengan al menos una palabra.

In [ ]:
def get_sentences(texto):
  ## Implementar aquí
  pass
  ##

In [ ]:
oraciones_limpias = get_sentences(texto)
oraciones_limpias

['[Letra de "¡Oh, Algoritmo!" ft. Nora Erez]',
 '[Refrán: Jorge Drexler]',
 '¿Quién quiere que yo quiera lo que creo que quiero?',
 '¿Quién quiere que yo quiera lo que creo que quiero?',
 '¿Quién quiere que yo quiera lo que creo que quiero?',
 '¿Quién quiere que yo quiera lo que creo que quiero?',
 '¿Quién quiere que yo quiera lo que creo que quiero?',
 '¿Quién quiere que yo quiera lo que creo que quiero?',
 '[Estribillo: Jorge Drexler]',
 'Dime qué debo cantar',
 'Oh, algoritmo',
 'Sé que lo sabes mejor',
 'Incluso que yo mismo',
 '[Verso 1: Nora Erez]',
 "Wait, what's that money that you spent?",
 "What's that sitting on your plate?",
 "Do you want what you've been fed?",
 'Are you the fish or bait?',
 "Mmm, I'm on the top of the roof and I feel like a jail",
 'Rather not pay the bail',
 'To dangerous people with blood on their faces',
 "So I'm sharing a cell with the masses",
 'The underground always strive for the main',
 "Streaming like Grande's big-ass ring",
 "Screaming: I'll wr

Debería obtener en total 87 oraciones.

In [ ]:
len(oraciones_limpias) == 87

87

Ahora definiremos el conjunto de entrenamiento y prueba para las oraciones:

In [ ]:
split = int(len(oraciones_limpias) * 0.8)
train_corpus = oraciones_limpias[:split]
test_corpus = oraciones_limpias[split:]

len(train_corpus), len(test_corpus)

(69, 18)

### 6.b Estimación de N-gramas (0.5 puntos)

Defina una función que reciba una lista de oraciones de un corpus y un N que indique el tamaño de los N-gramas. La función debe retornar un diccionario de Python donde la llave es un token (o palabra) y el valor es la cantidad de veces que ocurre el token, es decir, la frecuencia. En el caso de N-gramas con N mayor a 1 (como bi-gramas o tri-gramas) debe añadir un token especial al inicio o final de cada oración según corresponda (ver clases del curso).

In [ ]:
from nltk.tokenize import word_tokenize

In [ ]:
def n_grams(corpus, n=3):
  ## Implementar aquí
  pass
  ##

In [ ]:
n_grams(train_corpus, n=1)

In [ ]:
n_grams(train_corpus, n=2)

In [ ]:
n_grams(train_corpus, n=3)

Debe mostrar que su método funciona para $N = 1,2,3$.

## P7. Perplexity (1 punto)

En esta sección evaluarán su modelo de n-gramas y determinarán la probabilidad de oraciones y la perplejidad con un conjunto de test. Recuerde que la perplejidad se define de la siguiente manera:

$$
\text{Perplexity} = 2^{-l} \quad \quad l = \frac{1}{M} \sum_{i=1}^{m} \log p(s_i)
$$

con $m$ el número de oraciones del corpus y $M$ el tamaño del vocabulario.

### 7.a Obtener probabilidades (0.5 puntos)

En esta sección implementará una función que determine la probabilidad de una oración.

Defina una función que reciba una oración, un diccionario con n-gramas y el valor de $n$. La función debe entregar la probabilidad de cualquier oración.

**Hint**: No olvide los posibles casos borde, como palabras fuera del vocabulario.

In [ ]:
def get_probability(sentence, n_grams_frequency, n):
  ## Implementar aquí
  pass
  ##

Pruebe su función con oraciones frecuentes y comente sus resultados



### 7.b Perplexity en conjunto de test (0.5 puntos)

En esta sub-sección deberá calcular la perplejidad del corpus de test.

Defina una función que reciba un corpus de test y retorne la perplexity (ver clases del curso). Utilice la función de la sección anterior.

In [ ]:
def get_perplexity(corpus, n):
  ## implementar aquí
  pass
  ##

In [ ]:
get_perplexity(test_corpus, 3)

Dé una interpretacion de la perplexity en el corpus de test:
```
Nosotros ...
```

## P8. Interpolación Lineal (0.5 puntos)

Cree una función que obtenga la probabilidad de una oración interpolando linealmente modelos de unigrama, bigrama y trigrama ponderados por $\lambda_1, \lambda_2$ y $\lambda_3$ respectivamente. Para esto use las funciones que creó anteriormente.

In [ ]:
def get_probability_lineal_interpol(sentence, corpus, l_1, l_2, l_3):
  ## Implementar aquí
  pass
  ##

Defina una función para calcular la perplejidad de un corpus con interpolación lineal.

In [ ]:
def get_pp_interpol(corpus, l_1, l_2, l_3):
  ## Implementar aquí
  pass
  ##

Ahora haga pruebas con distintos valores de $\lambda_1, \lambda_2$ y $\lambda_3$, incluyendo valores extremos (por ejemplo $[1, 0, 0]$). Comente sus resultados.